kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp

```diff
@@ -0,0 +1,213 @@
+// Concord
+//
+// Copyright (c) 2018-2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include <pthread.h>
+#include <rocksdb/filter_policy.h>
+#include <rocksdb/table.h>
+
+#include "migration_bookeeper.hpp"
+
+using namespace boost::program_options;
+using namespace concord::storage::rocksdb;
+using namespace concord::util;
+
+int BookKeeper::migrate(int argc, char* argv[]) {
+  if (!parse_args(argc, argv)) {
+    return EXIT_FAILURE;
+  }
+
+  initialize_migration();
+
+  LOG_INFO(logger_, "Starting the migration");
+  if (config_["migrate-to-v4"].as<bool>()) {
+    migrate(cat_kvbc_->getAdapter(), v4_kvbc_->getAdapter());
+  } else {
+    migrate(v4_kvbc_->getAdapter(), cat_kvbc_->getAdapter());
+  }
+  LOG_INFO(logger_, "Successfully completed Migration, thanks for your time !!!");
+  return EXIT_SUCCESS;
+}
+
+void BookKeeper::initialize_migration() {
+  const auto read_only = true;
+  const auto migrate_to_v4 = config_["migrate-to-v4"].as<bool>();
+
+  auto rodb =
+      NativeClient::newClient(config_["rocksdb-path"].as<std::string>(), read_only, NativeClient::DefaultOptions{});
+  auto wodb =
+      NativeClient::newClient(config_["rocksdb-path"].as<std::string>(), !read_only, NativeClient::DefaultOptions{});
+  v4_kvbc_ = std::make_shared<BlockchainAdapter<concord::kvbc::V4_BLOCKCHAIN>>(
+      migrate_to_v4 ? wodb : rodb, false, migrate_to_v4);
+  cat_kvbc_ = std::make_shared<BlockchainAdapter<concord::kvbc::CATEGORIZED_BLOCKCHAIN>>(
+      migrate_to_v4 ? rodb : wodb, false, !migrate_to_v4);
+
+  storage_thread_ = std::make_unique<std::thread>([this, migrate_to_v4]() {
+    if (migrate_to_v4) {
+      store_to(v4_kvbc_->getAdapter());
+    } else {
+      store_to(cat_kvbc_->getAdapter());
+    }
+  });
+  if (std::thread::hardware_concurrency() > 2) {
+    cpu_set_t cpu_set;
+    CPU_ZERO(&cpu_set);
+    CPU_SET(1, &cpu_set);
+    int rc = pthread_setaffinity_np(storage_thread_->native_handle(), sizeof(cpu_set_t), &cpu_set);
+    if (rc != 0) {
+      LOG_WARN(logger_, "Cannot create a CPU affinity for storage thread, so this process might run slow.");
+    }
+  }
+  const auto point_lookup_threads = config_["point-lookup-threads"].as<std::int64_t>();
+  batch_thread_pool_ = std::make_unique<ThreadPool>(static_cast<std::uint32_t>(point_lookup_threads));
+}
+
+void BookKeeper::migrate(const std::shared_ptr<concord::kvbc::adapter::ReplicaBlockchain>& from,
+                         std::shared_ptr<concord::kvbc::adapter::ReplicaBlockchain>& to) {
+  start_batch_id_ = from->getGenesisBlockId();
```

Comment:
if the genesis block is greater than one, we'll need to perform a special pruning handling.